Fixes behavior of stc.save() as mentioned in issue #13158

shresth-keshari opened the pull request:
Fixes #13158 .

Doing stc.save("test.h5") saves a single test.h5 file, and doing stc.save("test.stc") saves to the two files test-lh.stc and test-rh.stc.

Please let me know if any more changes required in the current iteration.

drammock commented on mne/source_estimate.py:
```diff
@@ -1894,18 +1894,26 @@ def save(self, fname, ftype="stc", *, overwrite=False, verbose=None):
+                ftype = "w"
+            elif fname.endswith(".h5"):
+                ftype = "h5"
+            else:
+                raise ValueError("Cannot infer file type, please specify ftype.")
```
the suggestion in the linked issue was to fall back to `"stc"` if the file type could not be inferred (not raise an error).

drammock replied:
```suggestion
            else:
                logger.info(
                    "Cannot infer file type from `fname`; falling back to `.stc` format"
                )
                ftype = "stc"
```

drammock commented on mne/source_estimate.py:
```diff
@@ -1894,18 +1894,26 @@ def save(self, fname, ftype="stc", *, overwrite=False, verbose=None):
-            File format to use. Allowed values are ``"stc"`` (default),
-            ``"w"``, and ``"h5"``. The ``"w"`` format only supports a single
-            time point.
+        ftype : str | None
```
the suggestion in the linked issue was to use `"auto"` (not `None`) for the "infer-from-fname" behavior. I don't have very strong feelings one way or another, but @larsoner might?

drammock replied:
you changed the logic below, but didn't change this line.
```suggestion
        ftype : "auto" | "stc" | "w" | "h5"
```

drammock reviewed (commented):
CI errors look related:

https://github.com/mne-tools/mne-python/actions/runs/13924372063/job/38964972043?pr=13165#step:16:6285

> FAILED mne/tests/test_source_estimate.py::test_io_stc - OSError: SourceEstimate File(s) not found for: PosixPath('/private/var/folders/0j/bwqcs4y508s2n4ck4dhf3rpc0000gn/T/pytest-of-runner/pytest-0/test_io_stc0/tmp.stc')

Does mne/tests/test_source_estimate.py::test_io_stc pass for you locally?

shresth-keshari commented:
@drammock I have changed ftype from None to auto coz it was fairly easy. 
I also put it to test by saving a sample of both .stc as well as .h5 filetypes. You can see them in the screenshot, that they're being saved in the directory, named as it was suggested by @larsoner in the issue. (will be deleting these files before commit)

![Screenshot from 2025-03-18 22-39-54](https://github.com/user-attachments/assets/a552fc01-a555-4959-9aad-bb8dd4e35df9)

But, while running tests, it shows the following error:


![image](https://github.com/user-attachments/assets/9de60036-f529-4d5c-98d9-345e8d7bdf8a)

which says: FAILED test_source_estimate.py::test_io_stc - OSError: SourceEstimate File(s) not found for: PosixPath('/tmp/pytest-of-shresth/pytest-1/test_io_stc0/tmp.stc')

I'm not aware about how to fix this error. Please help.

hoechenberger commented on mne/source_estimate.py:
```diff
@@ -1884,7 +1884,7 @@ class SourceEstimate(_BaseSurfaceSourceEstimate):
 
     @verbose
-    def save(self, fname, ftype="stc", *, overwrite=False, verbose=None):
+    def save(self, fname, ftype="auto", overwrite=False, verbose=None):
```
Please don't remove the `*`

shresth-keshari replied:
Yeah, thanks for pointing out,
I had noticed that a couple of days back, added "*" as suggested, but still, when I run pytest, it fails with the same message as mentioned above.
Well, if I test the function using a sample stc file, it works perfectly.
I guess the problem is not limited only to this, and needs some other tiny fix.

shresth-keshari replied:
sure

drammock commented on mne/source_estimate.py:
```diff
@@ -1894,18 +1894,26 @@ def save(self, fname, ftype="stc", *, overwrite=False, verbose=None):
-            ``"w"``, and ``"h5"``. The ``"w"`` format only supports a single
-            time point.
+        ftype : str | None
+            File format to use. If "auto", the file format is inferred from the
+            file extension. Allowed values are ``"stc"``, ``"w"``, and ``"h5"``.
+            The ``"w"`` format only supports a single time point.
```
```suggestion
            File format to use. If "auto", the file format will be inferred from the
            file extension if possible. Other allowed values are ``"stc"``, ``"w"``, and
            ``"h5"``. The ``"w"`` format only supports a single time point.
```

drammock reviewed (commented):
> I had noticed that a couple of days back, added "*" as suggested, but still, when I run pytest, it fails with the same message as mentioned above.

The `*` is not related to the pytest failures

> Well, if I test the function using a sample stc file, it works perfectly.

Our tests often check for "corner cases" that are probably different than the approach you're trying.  I suggest running pytest with the `--pdb` flag so you can examine what the problematic values are when the test fails, and see if that helps you figure out what else needs to be changed.

shresth-keshari commented:
Hey, @drammock, I implemented your suggestions, and ran pytest with --pdb flag.
The issue seems to be coming from "read_source_estimate" in source_estimate.py, which looks like this in the terminal:

```
test_source_estimate.py:484: in test_io_stc
    stc2 = read_source_estimate(tmp_path / "tmp.stc")

..\source_estimate.py:330: in read_source_estimate
    raise OSError(f"SourceEstimate File(s) not found for: {fname_arg!r}")

E   OSError: SourceEstimate File(s) not found for: WindowsPath('C:/Users/kesha/AppData/Local/Temp/pytest-of-kesha/pytest-3/test_io_stc0/tmp.stc')
```

Now that I look at `test_source_estimate.py`, in the function definition of `test_io_stc`, the stc file is saved first, using stc.save() function. Later, the same file is being read using `read _source_estimate`. Since we changed the naming of the files being saved, I believe that's what causes the issue. Will try fixing that and will get back to you with a solution soon.



have attached screenshots as well, please have a look:


![image](https://github.com/user-attachments/assets/0235f1a7-2990-4ad6-b0a9-2f46a32773bc)
![image](https://github.com/user-attachments/assets/e8afe89e-51e6-4e65-bb87-97232bfdda37)

drammock commented on mne/source_estimate.py:
```diff
@@ -1894,18 +1894,29 @@ def save(self, fname, ftype="stc", *, overwrite=False, verbose=None):
         """
         fname = str(_check_fname(fname=fname, overwrite=True))  # checked below
+        if ftype == "auto":
+            if fname.endswith((".stc", "-lh.stc", "-rh.stc")):
+                ftype = "stc"
+            elif fname.endswith((".w", "-lh.w", "-rh.w")):
```
This can be simplified, as the first element in each `endswith` tuple already captures all cases we care about.

drammock commented on mne/tests/test_source_estimate.py:
```diff
@@ -480,9 +480,8 @@ def attempt_assignment(stc, attr, val):
-    stc.save(tmp_path / "tmp.stc")
-    stc2 = read_source_estimate(tmp_path / "tmp.stc")
-
+    stc.save(tmp_path / "tmp.stc", overwrite=True)
```
is `overwrite=True` really needed here?  I wouldn't have expected it to be, since it's the first `.save` action in the test.

drammock commented on sample.py:
remove this file

drammock commented on test.h5:
remove this file

drammock commented on doc/changes/devel/13165.bugfix.rst:
```diff
@@ -0,0 +1,4 @@
+Fixes behavior of stc.save() as mentioned in issue #13158 :meth:`stc.save` and in `test_source_estimate.py` by `Shresth Keshari`.
+- Changed filetype to "auto"
+- saving a `.stc` file automatically saves two files with suffix `-lh.stc` and `-rh.stc`, which is intuitive and removes any confusion.
+- there were problems with pytest later on, which were tackled by minor tweaks.
```
- simplify this to a single sentence, we don't normally include detailed bullet-lists.
- This is the source of the doc build errors, because what comes after `:meth:` needs to be a fully qualified path to the method (e.g., mne.SourceEstimate.save).
- remove the reference to the test file

shresth-keshari commented:
Hello, @drammock and @larsoner
I believe that most of the changes have passed checks, but after numerous pytests, I can't pinpoint where does the issue lie that is stopping the remaining checks from passing.

Please guide me ahead..

larsoner commented:
I merged `main` into your branch to see if it fixes the CIs
